[86]: Adding milvus as a vector database and a search service (RAG)

SatyamMattoo opened the pull request:
## Short description
Integrated a vector database embedded with the OpenFn docs and a search service for RAG.

## Related Issues
Fixes: #71 
Closes: #72 

## Implementation details
Added Milvus as the vector database, embedding the OpenFn docs from the [repo](https://github.com/OpenFn/docs) and adding it to Zilliz Cloud Database. You need to sign in to [Zilliz](https://cloud.zilliz.com/) and create a cluster to obtain the token and connection URI. 

The database can be seeded with the docs during the docker build using the command:
`docker build --secret id=_env,src=.env -t apollo .`

It can also be done by running the py script using:
`poetry run python services/search/generate_docs_emebeddings.py tmp/repo/docs/jobs collection_name` 
You will need to change the docs_path by the actual path to the docs stored on your machine i.e. `./tmp/repo/docs`. 


Add the OpenAI API key, Zilliz database URI, and token to the .env file. This mounts the environment variables to the command and helps protect the secrets as well.

The search service currently takes a search string and a OpenAI API key, embeds the search query, searches the vector database for relevant information, and returns the results to the CLI.

## Further work 
After the integration and embedding of the doc's repo into the vector database we will need to automate the docker build process, i.e. whenever any change is pushed to the main branch of the doc's repo, we will need to rebuild the docker image of the apollo repository. This might be achieved using GitHub actions and workflows:

1. Create a repository dispatch event as suggested in this [issue](https://github.com/OpenFn/docs/issues/520).
2. Listen to the dispatch event in the apollo repo and find a way to handle the env variables required for the docker build. (OpenAI API key, Milvus token and database URI)

## Acceptence Critiera

Before we can consider releasing this work, we need to:

(Added by @josephjclark )

* [ ] Settle on the embedding database and models. Should we try and build it all into the image or use a third party? If we use a third party, is Zilliz appropriate?
* [ ] Ensure the returned chunks are reasonable quality and sufficiently related to the query
* [ ] Decide whether to keep the summarization feature
* [ ] Work out whether and how to handle adaptor docs chunking

josephjclark commented:
Hi @SatyamMattoo, your mentor should be helping you with the review at this stage. We're going to get in touch to make sure the right procedure is being followed and that the mentor is on the right page.

josephjclark commented:
My guidance: I'd really like to avoid using a third party service if we can. Can we build a very small model into the image, rather than all of torch?  That was my expectation. If we are going to use third party services, we might as well do ALL of it with 3rd party services and build nothing into the model. 

If we really do have to build a lot of large models into the image, I need to see the evidence for it.

SatyamMattoo commented:
Hello @josephjclark,

I understand your concern regarding the use of third-party services for embedding models. After thoroughly researching various models, I found that most of them rely on torch as a dependency. Hence, building a small model directly into the image seems challenging.

However, here are a few approaches we can consider:

## 1. Adding Sentence-Transformers as a Dependency:

This approach significantly increases the build time up to 900 seconds.

![Screenshot 2024-07-09 182027](https://github.com/OpenFn/apollo/assets/96661612/607d96ba-9000-4281-bfb9-84220b41d5d1)


The resulting image size is approximately 13GB. Although this was my initial approach, the extended build time and large image size pose significant challenges. Moreover, embedding the actual documents will further increases the build time and image size. Therefore, directly including a model in the Docker image may not be the optimal solution.

## 2. Using OpenAI Model Provided by Milvus:

Instead of utilizing the Hugging Face API, we can leverage the OpenAI model provided by Milvus for embedding the documents. This approach also requires an API key. Since we are already using OpenAI keys, this might be a better alternative. You can find the documentation for this approach [here](https://milvus.io/docs/embed-with-openai.md). Additionally, for search queries, we can use the API key provided by the user to obtain the search embeddings.

## 3. Mounting the Model as a Volume:

We can mount our model using volumes. While I am uncertain about its viability in a production environment, this approach was mentioned [here](https://github.com/UKPLab/sentence-transformers/issues/352#issuecomment-671941477). This might work well locally, but its effectiveness in production remains unclear.

These are the approaches I have identified after extensive research over the past week. I would appreciate your thoughts on this. Am I missing anything here, or can you direct me to any resources that might offer a better approach?

Thank you!

josephjclark commented:
Let's break the problem into two.

First, we want to take a text corupus, and convert it into embeddings. This is an offline process outside the docker image (or in a builder image). Then in our docket file, we add an embeddings database and loading the trained data.

I don't expect the database or the embeddings to need torch. It's just a vector database and a bunch of data. 

I don't think we need torch and ml to search the database either. 

So we should be able to build this into an image without torch and a relatively small file size.

If either of these assumptions is wrong, I need to know so we can pivot.

Secondly, we need to convert natural language queries into  embeddings in the fly. This is the harder bit and this either needs an embeddings service or a torch-enabled image. But let's deal with this bit after we've conquered step one.

SatyamMattoo commented:
Thank you for your assistance sir. There are some details that I believe we need to discuss. Could we schedule a meeting to go over this and get a clearer idea?

SatyamMattoo commented:
@josephjclark Yes sir, I have. I will implement that for now and continue to look for a better approach to do it within the image in parallel.

josephjclark commented:
Some more thoughts on direction here:

### PR Description

I think the PR description has fallen out of sync with the latest commits: please keep the PR description up to date so that when someone from the team comes to do a  review, it's easy to get up to speed. Changes can be logged in comments (or commits) as you go, but the description should always reflect the current state of the work

### Hosted Embeddings

I was thinking, if we ARE going to use third party services for embeddings, how and when do those embeddings get updated?

The original design asked for the database to be built from scratch each time. But with a third party, when we build, the docs embeddings may well already exist in the third party database.

In fact, ideally the embeddings would be updated whenever the doc site is updated, and then the live server is always querying against the very latest version of the docs.

Also, when the third party database is updated, what happens to the old embeddings? Are they purged or are we paying to store old data? 

I don't necessarily need to see an implementation in this PR to automatically update the embeddings, but I would like to see a proposal for a robust, scalable strategy.

### Testing

What sort of testing can we use to ensure that the embeddings are working? How do we know the right content is embedded from the doc site? I don't necessarily mean unit testing. Manual tests or logging during the build process could be sufficient. What I'm looking for is an answer to the question "How do I know this is working?"

SatyamMattoo commented:
Sure sir, I will update the PR description shortly. 

For the updating the database we can set up GitHub actions and set the workflows inside the apollo and docs repositories to rebuild the docker image each time a change is pushed to the main branch of the docs repository.

### Steps:
1. A change is pushed to the main branch of the doc's repo, a GitHub Action workflow is triggered that rebuilds the docker image in the apollo repository.
    - Create a GitHub Actions workflow in the docs repository that triggers a repository dispatch event to the apollo repository 
      whenever there is a push to the main branch.
    - Create a GitHub Actions workflow in the apollo repository that listens for the repository dispatch event and rebuilds the 
      Docker image.

2. Rebuilding Docker Image:
     - The workflow in the apollo repo will clone the necessary folders from the docs repo.
     - It will then process the markdown files, chunk them appropriately, and embed the text using the embedding model.

3. Database Update:
    - Connect to the vector database. Check if the collection exists. If it does, drop the existing collection to remove old 
       embeddings.
    - Insert the new embeddings into the database.

4. Managing Embeddings:
    - By dropping the old collection and inserting new embeddings, we ensure that only the latest data is stored. This approach keeps storage costs efficient and maintains the relevance of the data.

### Testing:
For testing manually,
- I will add some more detailed logs during the docker build.
- We can actually see the text embedded into the vector database, here is a screenshot of the docs embedded into zilliz 
   
![Screenshot 2024-07-16 004713](https://github.com/user-attachments/assets/f367f5b1-7794-4ddb-917c-40671f7d7d47)

- Use the search service to query the database and ensure that it returns accurate results based on the latest embeddings. This 
   can be done by querying specific known sections of the docs and verifying the results.

josephjclark commented:
That's really cool that we can see the embeddings in the hosted database. I'll have to explore that.

I this stage I don't want to curate the docs content too heavily - I'm happy just to spam it all in there and come back and refine it later.

Thank you for the notes on automation. I don't particularly want to automate the docs repo or this repo right now, I just want you to propose a plan. Perhaps you can make a note in the PR under a section called Further Work, and maybe raise an issue here and in the docs repo to cover the work required.

What I DO want to see on this branch is the logic to drop the old embeddings on build

Remember the point of this work is to benefit the job expression service. So very soon you should branch off of this work and start integrating the search service with the job writing service. Then, I hope, we'll start seeing much improved job code generation.

Soon after that, we'll have to work out a demo or test suite or gallery or something that you can show us what the code output looks like. I have a few ideas but I don't know exactly what that will look like: I suggest that you think about an easy way to demo your work and propose something to us. I'm sure your mentor can help there.

SatyamMattoo commented:
Sure, sir. I will be integrating the search service to add context to the prompt in the job generation service. For that, I believe I will need to merge the dmp/620 branch with this one?
I have already added the logic to drop the old embeddings from the hosted database. This is included in the `init_milvus.py` file:

```
    # Check if the collection exists
    collection_name = "apollo_sample"
    check_collection = utility.has_collection(collection_name)
    if check_collection:
        drop_result = utility.drop_collection(collection_name)
```

Additionally, I have implemented another approach for chunking the documents using a text splitter. I have added a recursive text splitter along with the markdown splitter to avoid uneven chunks of data. Evenly sized chunks with overlap help us improve the search results.

Best regards.

josephjclark commented:
@SatyamMattoo When I try to run `init_milvus.py` from my file system (I've changed the path to the docs repo), I get this error:

```
openai.BadRequestError: Error code: 400 - {'error': {'message': "'$.input' is invalid. Please check the API reference: https://platform.openai.com/docs/api-reference.", 'type': 'invalid_request_error', 'param': None, 'code': None}}
```

Any ideas? The corpus looks to be a large array of strings, and the API token does seem to be set

josephjclark commented:
@SatyamMattoo I see - that seems to be working! The path needs to point to a repo containing docs and adaptors.

That's a surprise to me: I don't think we need to embed the adaptors repo? That's huge and I've already created a service to pull down the typescript definitions for the adaptor, which I think will suffice

SatyamMattoo commented:
@josephjclark, Yes I realized that, and I had removed the adaptors folder from the docker build, I will remove it from this file in the next commit. For the adaptor I have added the describe_adaptor service to add the context.

josephjclark reviewed (changes requested):
I am likely to want to start experimenting with this over the next couple of weeks - so we've got a good opportunity now to sharpen up this work.

First of all, I think you should be spending as much time as possible on AI problems, no infrastructure ones. That means getting the RAG stuff really good.

Before we can do that though, I am going to ask you for a little bit more organisation and infrastructure. I hope this won't take long to set up.

This PR is currently addressing two distinct projects: `search_docs` and `gen_job`.

I think we need to break this back up into two different branches, one for each project. `gen_job` needs to branch off of `search_docs`, and you'll need to merge them together occasionally. But please keep two PRs open. The search one targets main, the job code one targets the search one.

This should be easy to do: branch this into `search_docs` and delete the `gen_job` folder. Then  branch off into `gen_job`, and do `get checkout milvus-integration services/gen_job`. That' "pull" the latest job generator service into that branch.

I've included a bunch of review comments here: some simple tidyups which won't take you long, plus some deeper engineering issues which we need to work on over the next few weeks.

On the search side, I want to focus on visibility and debugging. I want to better see what we're embedding, and what embeddings we're passing to job generator.

On the job generator side, I want to be able to easily see samples of code output so that I can assess how well the system is performing (that's the point of all this, of course).

SatyamMattoo commented on Dockerfile:
```diff
@@ -2,9 +2,21 @@ FROM python:3.11-bullseye
 
+RUN apt-get update && apt-get install -y git
+
+RUN git init /app/repo
```
`git clone --depth 1 https://github.com/OpenFn/docs.git /app/repo` seems to do the job but it clones the entire repository. But since this only pulls the latest commit, the size is reduced significantly.
On the other hand, using sparse checkout we can just pull the docs folder from the repo.

The sparse checkout seems to take more memory. So, I guess we can use git clone.
Thanks for the suggestion.

SatyamMattoo commented on init_milvus.py:
```diff
@@ -0,0 +1,110 @@
+    connections.connect("default", uri=milvus_uri, token=token, db_name="apollo_db")
+
+    # Check if the collection exists
+    collection_name = "apollo_sample"
```
Yes, we can pass this as an argument to the script (I have done this for the docs_path). But passing it as an argument does not seem much useful at the moment. What do you think?

josephjclark replied:
I think it will be useful later.

Please declare this and any other CLI args the top of the file so that users a can work out the expected usage. It's probably worth a line of docs or two in the readme to explain how to manually run this file.

SatyamMattoo replied:
Sure, I will do it.

SatyamMattoo commented on init_milvus.py:
```diff
@@ -0,0 +1,110 @@
+
+    # Read markdown files from the repo
+    repo_path = "/app/repo"
+    docs_folders = ["docs", "adaptors"]
```
Sure, this will be more convenient. We can start off with the jobs folder for now.

SatyamMattoo commented on init_milvus.py:
```diff
@@ -0,0 +1,110 @@
+            docs.append(f.read())
+    return docs
+
+def split_md_by_sections(content):
```
Yes, the Zilliz UI does not provide a complete view of the embeddings. I have added a functionality where the file name and the section start of each split is logged to stdout. The sections are written to the tmp/sections_split folder along with the file name and the section start as their headers so that we can easily see how our data is being split. For now, the data is being split into 161 sections.

SatyamMattoo commented on services/gen_job/gen_job.py:
```diff
@@ -0,0 +1,30 @@
+    data = Payload(dataDict)
+    logger.info("Running job expression generator with gpt3_turbo model")
+
+    result = generate("gpt3_turbo", data.api_key, data.adaptor, data.instruction, data.state, data.expression)
```
Sure, for now the output for:
```
{
"existing_expression": "",
   "adaptor": "@openfn/language-dhis2@4.0.3",
   "state":" { 'data': { 'name': 'bukayo saka', 'gender': 'male' } }",
   "instruction": "Create a new trackedEntityInstance 'person' in dhis2 for the 'dWOAzMcK2Wt' orgUnit."
}
```

is returned as:

![Screenshot 2024-08-08 203216](https://github.com/user-attachments/assets/d0d5d5bd-7df9-428d-84ed-93dadc8f37ab)

This is quite close to what is expected but there is still scope for improvement. What do you think about this?

SatyamMattoo replied:
Yes, there is a lot of scope for improvement here. We can improve the code quality by improving the RAG service but how do we plan to add the attribute IDs to the prompt? 
Is this something we can expect:
![Screenshot 2024-08-11 214640](https://github.com/user-attachments/assets/2fa87db1-19ab-4e60-86de-62aefbc82b9e)
 
Also, what if we plan to deploy a fine-tuned model to OpenAI and call that model? This would increase the quality of response significantly.

josephjclark replied:
Attribute IDs are an interesting problem to solve.

We can't add DHIS2 specific rules to the prompt. Maybe we can add some prompt rules to use placeholders if it doesn't know real values.

We do have some helper APIs to get lists of attribute IDs for a given dhis2 server... but that's a lot of work to set up, and it still gives us a dhis2 specific prompt.

I think placeholders, like in your screenshot, are the best we can hope for at the moment

josephjclark replied:
Fine tuning our own model is something we want to explore but is outside the scope of this project.

josephjclark commented:
@SatyamMattoo if you want to keep this branch and this PR as the `search` feature, you'll need to update the PR title and description.

josephjclark commented on services/search/generate_docs_emebddings.py:
```diff
@@ -0,0 +1,125 @@
+import os
```
This file name is mis-spelled

SatyamMattoo commented on services/search/generate_docs_emebddings.py:
```diff
@@ -0,0 +1,125 @@
+    os.makedirs(output_dir, exist_ok=True)
+    with open(os.path.join(output_dir, f"{os.path.basename(file_name)}_sections.md"), "w", encoding="utf-8") as out_file:
+        for section in splits:
+            out_file.write(f"## File: {file_name}\n")
```
It was really hard to differentiate between the chunks, therefore I added this. Since this does not seem to be useful, I will just add a heading as `"## Section Start: " + "\n---\n"` for each chunk. It will be much more readable, I guess?

SatyamMattoo commented on services/search/generate_docs_emebddings.py:
```diff
@@ -0,0 +1,125 @@
+    return docs
+
+def split_md_by_sections(file_name, content):
+    headers_to_split_on = [
```
The problem here is that if we just split the documents based on headings it would result in very uneven chunks (some greater than 2700). Therefore, using a text splitter after chunking based on headers seems necessary. 

The character splitter provides a parameter `chunk_overlap` which helps us to maintain context between the chunks.

This helps in maintaining proportionate chunk size and also maintains context between the chunks. Also, only sections with greater size are split with some overlap and we can change the value to see what works for us.
```
    text_splitter = RecursiveCharacterTextSplitter(chunk_size=512, chunk_overlap=128, length_function=len, 
    is_separator_regex=False)
    splits = text_splitter.split_documents(md_header_splits)
```

josephjclark replied:
Great, thank you for flagging this! Please do raise issues as you find them and bring out discussion points in prominent places.

I am shutting down for the weekend but a couple of thoughts:

- Is uneven chunk size really a problem? Is it just a problem of input token size?
- Can we add metadata to the documentation to aid chunking?
- Would it help to re-write the docs to have more even sections? I have no problem re-writing docs to aid ML (so long as it doesn't hinder human learning, of course)

If we need to experiment with the split parameters, I would encourage you to spend some time on this. Focus on the job writing guide. You should be able to look at the chunks and, as a human, feel like it's describing a meaningful unit of knowledge. A fixed set of concepts. If you look a chunk at random - can you tell what it's talking about?

Are there any other pre or processing steps that could help here? You may want to research this

SatyamMattoo replied:
- Yes, uneven chunking affects the search results and eventually the input token size.
- We can add metadata, but the overlap factor already adds context between the chunks.
- Yes, we can do that as well, but in future when we plan to embed all of the docs, this won't be a good approach, I guess. 
About the pre processing steps, I think it will be highly effective to deploy a fine tunned model trained on OpenFn datasets and call that model. Also, once this is implemented, we can use feedback loops to improve the results.

SatyamMattoo commented on services/search/generate_docs_emebddings.py:
```diff
@@ -0,0 +1,125 @@
+
+    # Log the first line of each section and the file name to stdout
+    for section in splits:
+        first_line = section.page_content.splitlines()[0] if section.page_content.splitlines() else "[Empty Section]"
```
Sure, I will add that

SatyamMattoo commented on services/gen_job/prompt.py:
```diff
@@ -0,0 +1,68 @@
+    query = f"Get the job writing guide, Usage Examples, and Job Code Examples. Instruction: {instruction}"
+
+    data_dict = {"query": query, "api_key": api_key}
+    search_results = apollo("search", data_dict)
```
I added this when I was embedding all of the docs into the database. Now I have removed this from the vector search. I have added more visualization to the context drawn from the database by writing the context and adaptor description to a file in the tmp folder.

About the metadata, yes we can add metadata to the database along with the chunks. We just need to add another field to the collection. What metadata can we add along with the chunks in that case?

josephjclark commented:
Totally random thought

When indexing the API reference, we need to create one embedding per fucntion. Those docs are very very highly structured and we should be able to create really accurate chunks.

Given a question like "how do I create an sobejct in salesforce", the search should return the API reference for the `create` function (and probably `upsert` too but you take my point)

SatyamMattoo commented:
I guess you are referring to the [adaptor](https://docs.openfn.org/adaptors) docs? Correct me if I'm wrong as I am not sure about what API reference here is. 

Yes, they are very well structured, and we can get even chunks just by splitting them based on headers and we might add some metadata related to the adaptor. Also, I have noticed the describe_adaptor information that is being added to the prompt, plays a major role in returning good results. I think it would be beneficial to embed the adaptor docs to the database and then retrieve information. Cause the describe adaptor service makes the input prompt quite large.

PS: I have some issues with my device that is taking a few days to fix, I will be back on this as soon as that gets fixed.

josephjclark commented:
Hey @SatyamMattoo, that comment wasn't neccessarily aimed at you - just a note. 

But yes, I am talking about the adaptor docs.

I'm not sure if this is a priority for your last few weeks on this, but it's certainly a candidate to explore. On refletion it's a particularly strong use-case for RAG.

josephjclark commented on services/search/generate_docs_emebeddings.py:
```diff
@@ -0,0 +1,129 @@
+import os
```
There is a typo in this file name.

Have you done  "self review" of this code? Has your mentor seen it and given any feedback?

SatyamMattoo replied:
Sorry, I missed this out. I will fix this in the next commit.

josephjclark commented:
@SatyamMattoo if you have time to work on DMP this week I would like you to investigate creating one rag chunk per adaptor function. You might find you need different rules for chunking the adaptors docs versus the main docs.

josephjclark commented on services/search/generate_docs_emebeddings.py:
```diff
@@ -0,0 +1,129 @@
+    )
+
+   # Read markdown files from the repo
+    docs_path = args.repo_path
```
Are we currently generating embedding for the _whole_ docs repo? That's going to generate a lot of noise. I thought were just indexing the job writing guide, or maybe just the docs subpath

Can you think of a good way to config drive this stuff? We take a path to the docs repo as an argument, but then we need to capture an array of paths _within_  the repo to parse.

I don't think we need to change this config file very often, but we do need the tracked paths to be easily discoverable. A config file called `paths.config` with a glob per line - like a gitignore - might be a good solution here? The file would need to support comments

And as mentioned elsewhere, I think we need to handle the adaptor embeddings a bit differently

SatyamMattoo replied:
No, we are not generating embeddings for the whole repo, it is just for the jobs folder currently. We are passing the path to embed as `tmp/repo/docs/jobs`. This works well for embedding a single folder. However, as we plan to embed more documentation in the future, I might need to rethink this approach.

One idea I had while working on this was to maintain a list of folder names that we need to embed, which can be adjusted as needed. For example: `const path = [ "jobs", "adaptors"]`.
What are your thoughts on this approach? I’ll implement the method you suggested today, and then we can compare both approaches to see which one is better in the long run.

About handling the chunking part, I have been working on exploring different methods. One of them is [semantic chunking](https://medium.com/the-ai-forum/semantic-chunking-for-rag-f4733025d5f5) which creates chunks based on similarity between the sentences. This seems to be quite effective. We can use this to chunk the job docs. 
For the adaptor docs we can simply use header splitters and add metadata. I will give updates after I have implemented and tested this.

Apart from this, I have been trying to implement a hybrid approach to retrieve relevant chunks. Milvus uses three search methods:

- Dense Retrieval: Utilizes semantic context to understand the meaning behind queries.
- Sparse Retrieval: Emphasizes keyword matching to find results based on specific terms, equivalent to full-text search.
- Hybrid Retrieval: Combines both Dense and Sparse approaches, capturing the full context and specific keywords for comprehensive search results.

We can read more about this [here](https://milvus.io/docs/hybrid_search_with_milvus.md)

These changes should help improve our embeddings and search results. I'll push the changes after I’ve tested them along with the other updates.

SatyamMattoo commented:
@josephjclark Sure, I am on it. Will be pushing the updates in a couple of days.

SatyamMattoo commented:
Hello @josephjclark,
I have pushed a few changes
1. Added semantic chunking method which is providing better quality chunks. We can the change the `breakpoint_threshold_amount` to increase or decrease the size of chunks (something around 80 seems fine).
2. There are 3 methods which we can use to add paths to embed. 
    - (used currently) Maintaining a list of file paths to be embedded and appending that to the argument provided
    - Adding all the paths as arguments. (Might not be efficient I guess?).
    - The one you suggested. It is quite similar to the first method, we just read the paths from a separate .config file.

About the hybrid search, it will not be possible as OpenAI embeddings do not support sparse embeddings. 

About the adaptor docs, I will need more time to figure that thing out. The docs repo generates the adaptor functions 
and they are not present in the md files (they are generated during the build if I am not wrong). Would running [this](https://github.com/OpenFn/docs/blob/main/generate-adaptors/index.js) file do the job? Another way might be creating a similar function and retrieving functions from [this](https://[raw.githubusercontent.com/OpenFn/adaptors/docs/docs/docs.json](https://raw.githubusercontent.com/OpenFn/adaptors/docs/docs/docs.json)) url? I will need some assistance on this.

SatyamMattoo commented:
Hello @josephjclark, I have added a feature to the search service where we can summarize the retrieved context based on the user query. I think this would really be beneficial and will return relevant and concise context based on the query. I have kept this as an optional field for now.

I have included options which will provide us more customization over the search. Now we can have these options in the input payload
```js
{
    "api_key": "<OpenAI api key>",
    "query": "What are jobs in OpenFn?",
    "limit": 10, // A custom limit for number of chunks retrieved ranging from 1 to 15
    "collection_name": "openfn_docs", // Might be helpful when we embed the adaptor docs to a new collection
    "summarize": true // Summarizes context based on your query
}
```
About the semantic chunking, for now it does not provide an option to limit max_tokens in a chunk. It will be added later but for now we can create a custom class based on the SemanticChunker. [Here](https://github.com/langchain-ai/langchain/discussions/22079) is more information about this. Also, if the chunks are too short, we can add them to the previous or next chunk. This will help us get even better chunks.

josephjclark commented:
I'm not sure I agree with you on the semantic chunking. I suppose I don't have the old chunks to compare to, but I think the quality of the chunks looks pretty poor.

For one thing they look very uneven: in `job-examples.md_sections.md`, there are several examples in one big chunk (looks like 5800 characters).  Then there's a chunk in the job writing guide that's just a single sentence.

Skimming over the chunks, I also see blocks of text which are clearly  (to me) related - like some example code and its caption - which are split over several chunks. I see chunks begin on a sentence which relates back to the previous sentence. I see splits across HTML elements where the content must surely all be closely related.

Here's a random example:
```
## Section Start: 
------
It cannot be used in place of the `state` variable. It cannot be assigned to, or
be on the left hand side of an assignment, and can only be used inside an
argument to a function

This also means that Lazy State Operator can only be used to READ from state. It
cannot be used to assign to state directly. These examples are all errors:

```js
❌ const url = $.data.url;
get(url);

❌ get(() => $.data.url);

❌ $.data.x = fn();

❌ fn(state => {
  $.data.x = 10;
});
```

<details>
<summary>Compilation rules for advanced users</summary>

How does the Lazy State Operator work?
```

If you gave me, as a human, several of these chunks, I wouldn't understand what you were talking about. I don't expect an AI to fare much better.

When chunking by section, I at least know that each chunk has a fair chance of having enough context for it to be useful by itself. I also know that we can re-write docs in such a way that each section is designed to be read standalone (that's just good practice anyway). I'm sure there are some simple heuristics we can use to break up the longer chunks too.

josephjclark commented on services/search/search.py:
```diff
@@ -0,0 +1,73 @@
+            if isinstance(hits, str):
+                hits = json.loads(hits)  # Parse string to list of dictionaries
+            for hit in hits:
+                documents.append(hit['entity']['text'])        
```
It might be useful to output the document at this point, if only as a temporary debugging thing. Because when I run locally, it's impossible to read the result. even manipulating the resulting json is a bit hard  because of the nature of the strings and all their lien breaks.

Logging the strings to stdout here would be very readable I think

josephjclark commented:
I've run a couple of basic tests and while I'm not sold on the utility of the chunks, it does appear at first glance like the service is actually returning relevant chunks. When I ask about lazy state for example, I get ten chunks of lazy state data.

Can we control the relevance threshold at all? For our purposes, we don't need the top ten results - we need _all results that are relevant enough_. All results that score above 0.8 or whatever. If the search decides there are no docs worth returning, that's just fine (as would be the case when the user asks "What is Madonna's best album?")

josephjclark commented:
I am concerned about the summarisation process. This feels like using AI for the sake of AI to me.

Consider the pipeline. When the user asks a question, we must:
* Encode the question into embeddings (via a third party service)
* Search for relevant documents (via a third party service)
* Add the documents to the prompt
* Send the prompt to the model (via at third party service)

The final model must then parse the input prompt and break it down into its own embeddings.

Each step has a cost attached (environmental, temporal, financial and technical). It's a lot of computation already.

And then the summarizer adds an extra step, calling out to another third party service. Assuming the summarisation is good and accurate (which I don't believe), what is the value? Users will never see it.

The idea of generating text in one model to pass it as input to another model is frankly a bit scary.

As a general point of feedback: experimental fringe features like this should really be developed on different branches. It is harder for us to release and maintain code when it contains features that haven't been properly specced or tested. If the work was done on another branch, it could be examined, tested, discussed and released in isolation. Now it's tied up with everything else and will slow down the release.

I'm always happy to sound out new ideas and try new things. But I'm less happy when the experiments threaten the release of the main feature. When they're done in isolation, I can explore the idea in a safe space without having to worry about how it'll impact the release.

Don't branch the work off now. I'll have a think about it. Maybe you can convinced me that it's worth keeping the feature.

josephjclark commented on services/search/generate_docs_embeddings.py:
```diff
@@ -0,0 +1,144 @@
+    )
+
+    # Combine the base path with the additional folders
+    docs_to_embed = ["docs/jobs"]
```
This list is hard-coded, so if we ever want to update the list of paths, we have to call a python expert in to find the right line of code.

This should really be in a config file, like I asked, with one path per line. Bonus points if this config file accepts standard glob patterns.

The config file needs to be fully documented in the readme so that  developers unfamiliar with the codebase can update the paths a year from now.

josephjclark commented:
@SatyamMattoo just a quick note that I've added a list of acceptance criteria to the main PR post. These are things we need to work out before we can release this feature. I'll need to decide on the final embedding strategy next week so don't worry about that bit.

josephjclark commented:
Regarding adaptor docs, you should be able to just run `yarn generate-adaptors` from the docs repo and it'll build the docs as markdown to `adaptors/packages`. You'll only need to read the `<adaptor>-docs.md` files. You should be able to chunk up by section.

SatyamMattoo commented:
Hello @josephjclark,

Yes, it seems that semantic chunking won’t be the best fit for our docs. I’ve reverted to the previous strategy we used, which involves splitting based on headers and recursively splitting larger chunks.

Regarding the search, I’ve implemented a parameter that filters vectors based on a threshold. We can read more about this [here](https://milvus.io/docs/single-vector-search.md#Range-search). This ensures that irrelevant search queries return no results, and it works as expected.

As for the adaptor docs, I’ve worked out a solution locally. However, during the Docker build, it becomes a bit cumbersome since we need to install nodejs, npm, corepack, and yarn (with nodejs and npm required for installing yarn). This process adds about two minutes to the build time. Afterward, we still need to build the docs, which would further increase the image size and build time.

Alternatively, we could fetch the docs directly from the [GitHub source](https://raw.githubusercontent.com/OpenFn/adaptors/docs/docs/docs.json). There’s likely a way to retrieve functions and examples for each adaptor, create a chunk for each function along with its adaptor name, and generate embeddings from them. I had initially started with this approach before your suggestion.

What are your thoughts on proceeding this way?

SatyamMattoo commented:
@josephjclark Regarding the summarization process, I believe it might be more beneficial to discuss this after the release, as it may not directly impact the job generation service. The summarization essentially condenses the context based on the query, which we can potentially handle by adding specific instructions to the job prompt.

However, this could be useful if the user wants to use the search service directly for complex queries, as those can be answered based on the retrieved documents. I have removed this part for now, but this is something we can consider further down the line.

Also, I’ll make sure to branch off such ideas separately. 

Best regards

josephjclark commented:
Thank you for the update @SatyamMattoo 

You must take where when submitting work to thoroughly test it (we've talked about this before). And ideally your test results should be shareable, so that I can check your testing. It's not enough to say "this works great". You must _show_ me it works.

If you cannot prove the quality and robustness of your solution, it will not be merged. That's not my personal policy for openfn - that's how software development works worldwide. Different organisations have different standards, admittedly, but no-one wants to merge untrusted code.

I say this because you told me that `semantic chunking method is providing better quality chunks`. But you did not provide any proof of this, and now you've had to revert the work.

Experimentation is great - I'm glad we tried semantic chunking (and actually I'm sure we could spend more time on it to find a better solution). But you MUST test well and share your results. Super important.

The search threshold sounds great. What value have you chosen and why? How you can prove to me that it's a reasonable value?

Re adaptors: yes, using docs.json from github is a good solution, let's go with that for now.

SatyamMattoo commented:
Hi @josephjclark,

Thank you for the valuable advice; I will definitely keep these points in mind.

Regarding the threshold value, I’ve conducted tests with multiple queries to determine an appropriate value that returns only valid results (in most cases). For example, when testing with an unrelated query like "What is Langchain?", setting the threshold slightly higher (e.g. 1.4) still returns some results for optional chaining.

For relevant queries, I’ve observed that the ideal score range is between 0.7 and 1.2 for the chunks related to the query. We can already see these scores in the stdout. Additionally, I’ve pushed a file containing 2 sample queries along with their respective scores for further reference.

SatyamMattoo commented:
Hi @josephjclark,

josephjclark commented:
Thank you for your efforts here @SatyamMattoo 

I have spun out a couple of issues for further work which we'll be starting in a couple of weeks. In the meantime I'm going to merge and release this to the staging server :tada:

SatyamMattoo commented:
Hello @josephjclark,
That’s great news! Thank you so much for your guidance and support throughout—it’s been a great learning experience for me.
